### ISSUE TITLE
List group with anchors and buttons semantic html and accessibility

### ISSUE THREAD

**Issue by mtscout6** [twbs/bootstrap#16204]
I'm reviewing pull request react-bootstrap/react-bootstrap#467 and could use some input on a clear direction to take the `ListGroup` and `ListGroupItem`  React components we are putting together there.

When working with a `list-group` component the docs highlight two major code paths using either `div` or `ul` which lead me to this example:

```
<div>
  <ul class="list-group">
    <li class="list-group-item">Dapibus ac facilisis in</li>
    <li class="list-group-item">Cras sit amet nibh libero</li>
    <li class="list-group-item">Porta ac consectetur ac</li>
    <li class="list-group-item">
      <a href='#'>Vestibulum at eros</a>
    </li>
  </ul>
  <div class="list-group">
    <a href="#" class="list-group-item">Dapibus ac facilisis in</a>
    <a href="#" class="list-group-item">Cras sit amet nibh libero</a>
    <a href="#" class="list-group-item">Porta ac consectetur ac</a>
    <span class="list-group-item">Porta ac consectetur ac</span>
    <button class="list-group-item">Vestibulum at eros</button>
  </div>
</div>
```

Which produces:

![](http://cl.ly/image/1a3D1N3P2l11/Image%202015-04-01%20at%209.11.42%20AM.png)

I thought about using a vertical button group, but the background colors for contextual items are darker with the btn-group.

Have you considered having buttons within a list-group styled the same as the anchors? Or on the flip side having anchors and buttons that are direct children of an `li` appear the same as the `div` approach?

**Comment by zacechola** [comment]
Prototype of this: http://jsbin.com/megayo/1/edit?html,css,output

**Comment by twbs-lmvtfy** [comment]
Hi @zacechola!

You appear to have posted a live example (http://jsbin.com/megayo/1/edit), which is always a good first step. However, according to [Bootlint](https://github.com/twbs/bootlint), **your example has some Bootstrap usage errors**, which might potentially be causing your issue:
* line 28, column 9: [W007](https://github.com/twbs/bootlint/wiki/W007): Found one or more `<button>`s missing a `type` attribute.

You'll need to **fix these errors** and post a revised example before we can proceed further.
Thanks!

(*Please note that this is a [fully automated](https://github.com/cvrebert/lmvtfy) comment.*)

**Comment by zacechola** [comment]
Of course I missed one of those. Forcing re-lint: http://jsbin.com/megayo/2/edit?html,css,output

**Comment by mtscout6** [comment]
That is the desired output I'm looking for, thanks for the jsbin example @zacechola. Would a PR be welcome for that addition?

**Comment by mtscout6** [comment]
Also, what about the scenario of an anchor or button inside an <li> tag?

**Comment by zacechola** [comment]
When you put elements within `.list-group-item`, you're entering [custom list group content](http://getbootstrap.com/components/#list-group-custom-content) territory.

As for the possibility of a PR, I'd like to get @mdo's opinion first.

**Comment by patrickhlauke** [comment]
purely from a semantic markup point of view (as well as an accessibility angle), I'd actually say that the examples that are *not* using ``<ul>``/``<li>`` are sub-optimal...

**Comment by zacechola** [comment]
We should note that in the documentation, @patrickhlauke. Both `<ul>` and `<div>` are used in the list group examples.

**Comment by mdo** [comment]
Accessibility aside, the component was built with flexibility in mind. Use `<a>`s, `<div>`s, or `<li>`s depending on the needs of your team and your users. `<button>`s are explicitly supported, but it looks like they'd need only a tab more work to do so.

**Comment by patrickhlauke** [comment]
On the specific example of consecutive ``<a>``s, the example in our docs would definitely benefit from actual list markup around it (in my day job, we'd flag that as a failure in an accessibility audit). Once you have larger blocks of content, ``<div>``s without actual list markup is acceptable, depending on the exact content. I'll make a note to update/clarify the documentation...

**Comment by mtscout6** [comment]
It's due to accessibility audits that I'm looking into this, what's the thought on supporting the same styling of `<div>`s with `<a>`s, and `<li>`s with `<a>`s/`<button>`s?

**Comment by mdo** [comment]
I'm fine with supporting the `<button>` elements here, too.
